dotnet/src/Plugins/Plugins.Memory/MemoryBuilder.cs

```diff
@@ -0,0 +1,122 @@
+using Microsoft.SemanticKernel.Http;
+using Microsoft.SemanticKernel.Memory;
+
+namespace Microsoft.SemanticKernel.Plugins.Memory;
+
+/// <summary>
+/// A builder for Memory plugin.
+/// </summary>
+public sealed class MemoryBuilder
+{
+    private Func<IMemoryStore>? _memoryStoreFactory = null;
+    private Func<ITextEmbeddingGeneration>? _embeddingGenerationFactory = null;
+    private IDelegatingHandlerFactory _httpHandlerFactory = NullHttpHandlerFactory.Instance;
+    private ILoggerFactory _loggerFactory = NullLoggerFactory.Instance;
+
+    /// <summary>
+    /// Build a new instance of <see cref="ISemanticTextMemory"/> using the settings passed so far.
+    /// </summary>
+    /// <returns>Instance of <see cref="ISemanticTextMemory"/>.</returns>
+    public ISemanticTextMemory Build()
+    {
+        var memoryStore = this._memoryStoreFactory?.Invoke() ??
+            throw new SKException($"{nameof(IMemoryStore)} dependency was not provided. Use {nameof(WithMemoryStore)} method.");
+
+        var embeddingGeneration = this._embeddingGenerationFactory?.Invoke() ??
+            throw new SKException($"{nameof(ITextEmbeddingGeneration)} dependency was not provided. Use {nameof(WithTextEmbeddingGeneration)} method.");
+
+        return new SemanticTextMemory(memoryStore, embeddingGeneration);
+    }
+
+    /// <summary>
+    /// Add a logger factory.
+    /// </summary>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
+    /// <returns>Updated Memory builder including the logger factory.</returns>
+    public MemoryBuilder WithLoggerFactory(ILoggerFactory loggerFactory)
+    {
+        Verify.NotNull(loggerFactory);
+        this._loggerFactory = loggerFactory;
+        return this;
+    }
+
+    /// <summary>
+    /// Add a http handler factory.
+    /// </summary>
+    /// <param name="httpHandlerFactory">Http handler factory to add.</param>
+    /// <returns>Updated Memory builder including the http handler factory.</returns>
+    public MemoryBuilder WithHttpHandlerFactory(IDelegatingHandlerFactory httpHandlerFactory)
+    {
+        Verify.NotNull(httpHandlerFactory);
+        this._httpHandlerFactory = httpHandlerFactory;
+        return this;
+    }
+
+    /// <summary>
+    /// Add memory store.
+    /// </summary>
+    /// <param name="store">Store to add.</param>
+    /// <returns>Updated Memory builder including the memory store.</returns>
+    public MemoryBuilder WithMemoryStore(IMemoryStore store)
+    {
+        Verify.NotNull(store);
+        this._memoryStoreFactory = () => store;
+        return this;
+    }
+
+    /// <summary>
+    /// Add memory store factory.
+    /// </summary>
+    /// <param name="factory">The store factory.</param>
+    /// <returns>Updated Memory builder including the memory store.</returns>
+    public MemoryBuilder WithMemoryStore<TStore>(Func<ILoggerFactory, TStore> factory) where TStore : IMemoryStore
```

Comment:
Should this be named `WithMemoryStoreFactory`?

Reply:
While we have `WithLoggerFactory(ILoggerFactory loggerFactory)`, where we pass `Factory` interface, I don't think it's the case for MemoryStore. The idea here is to create one `WithMemoryStore` method which will have 3 different overloads:
1. `IMemoryStore store` - pass MemoryStore instance as it is.
2. `Func<ILoggerFactory, TStore> factory` - factory for MemoryStore initialization with access to ILoggerFactory.
3. `Func<ILoggerFactory, IDelegatingHandlerFactory, TStore> factory` - factory for MemoryStore initialization with access to ILoggerFactory and IDelegatingHandlerFactory.

Not sure if different name for methods that accept delegate will provide some benefits.